source/common/common/inline_map.h

```diff
@@ -0,0 +1,460 @@
+      ASSERT(inline_keys_map_.size() < std::numeric_limits<uint32_t>::max());
+      Handle handle{static_cast<uint32_t>(inline_keys_map_.size())};
+      inline_keys_.push_back(StorageKey(key));
+      inline_keys_map_.emplace(key, handle);
+      return handle;
+    }
+  }
+
+  /**
+   * Fetch the handle for the given key. Return handle if the given key is inline key and return
+   * absl::nullopt if the given key is normal key. May only be called after finalize(). This should
+   * be used to get the handle of the inline keys that added by addHandle(). This function could
+   * used to determine if a key is added as inline key or not at runtime or xDS config loading time
+   * and decide if the key should be used as inline key or normal key.
+   * Heterogeneous lookup is supported here.
+   */
+  template <class LookupKey> absl::optional<Handle> getHandleByKey(const LookupKey& key) const {
+    ASSERT(finalized_, "Cannot get inline handle before finalize()");
+
+    if (auto it = inline_keys_map_.find(key); it == inline_keys_map_.end()) {
+      return absl::nullopt;
+    } else {
+      return it->second;
+    }
+  }
+
+  /**
+   * Get the inline keys map that contains all inline keys and their handle. May only be called
+   * after finalize().
+   */
+  const InlineKeysMap& inlineKeysMap() const {
+    ASSERT(finalized_, "Cannot fetch registration map before finalize()");
+    return inline_keys_map_;
+  }
+
+  /**
+   * Get the array that contains all added inline keys. May only be called after finalize().
+   */
+  const InlineKeys& inlineKeys() const {
+    ASSERT(finalized_, "Cannot fetch registration set before finalize()");
+    return inline_keys_;
+  }
+
+  /**
+   * Get the number of the inline keys in this descriptor. May only be called after finalize().
+   */
+  uint64_t inlineKeysNum() const {
+    ASSERT(finalized_, "Cannot fetch registration map before finalize()");
+    return inline_keys_map_.size();
+  }
+
+  /**
+   * Finalize this descriptor. No further changes are allowed after this point. This guaranteed that
+   * all map created by the process have the same variable size and custom inline key adding. This
+   * method should only be called once.
+   */
+  void finalize() {
+    ASSERT(!finalized_, "Cannot finalize() an already finalized descriptor");
+    finalized_ = true;
+  }
+
+  /**
+   * @return true if the descriptor is finalized.
+   */
+  bool finalized() const { return finalized_; }
+
+private:
+  bool finalized_{};
+  InlineKeys inline_keys_;
+  InlineKeysMap inline_keys_map_;
+};
+
+/**
+ * This is the inline map that could be used as an alternative of normal hash map to store the
+ * key/value pairs.
+ */
+template <class StorageKey, class StorageValue> class InlineMap : public InlineStorage {
+public:
+  using TypedInlineMapDescriptor = InlineMapDescriptor<StorageKey>;
+  using Handle = typename TypedInlineMapDescriptor::Handle;
+
+  using DynamicHashMap = absl::flat_hash_map<StorageKey, StorageValue>;
```

Comment:
I think DynamicHashMap and TypedInlineMapDescriptor can be in the private section.

Reply:
got it.